File: GRAPH/C++/GRAPHS USING OOPS/Undirected_Graph_Adjacency_List.cpp
#include<bits/stdc++.h>
using namespace std; 

void add_edge(vector<vector<int>> &mat, int i, int j){
    auto it = find(mat[i].begin(), mat[i].end(), j);
    if(it == mat[i].end()){
        mat[i].push_back(j);
        if(i != j)
        mat[j].push_back(i);
    }
};

void display_Adj_List(vector<vector<int>> &mat){
    int v = mat.size();
    for(int i=0; i<v; i++){
        if(!mat[i].empty()){
            cout<< i << " : ";
            for(auto j : mat[i])
                cout<< j << " ";
            cout<<endl;
        }
    }
};

int main(){
    int v = 4 ;
    vector<vector<int>> mat(v);
    
    add_edge(mat, 0, 2);
    add_edge(mat, 0, 2);
    add_edge(mat, 2, 2);
    add_edge(mat, 2, 3);
    
    display_Adj_List(mat);
}

File: GRAPH/PYTHON/WEIGHTED/Undirected_Graph_Adjacency_List.py
def add_edge(i, j, w):
    if i not in Graph:
        Graph[i] = []
    if j not in Graph[i]:
        Graph[i].append([j,w])
    if j not in Graph:
        Graph[j] = []
    if i not in Graph[j]:
        Graph[j].append([i,w])

def display_Adj_List():
    for key, values in Graph.items():
        print(f'{key} ------> {values}')

Graph = {}
add_edge(0, 1, 3)
# add_edge(1, 0, 3)
add_edge(0, 1, 6)
add_edge(0, 2, 7)
add_edge(1, 2, 7645)
add_edge(1, 2, 56)
add_edge(2, 0, 45)
display_Adj_List()